File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/procurement/goods-receipt/goods-receipt.model.ts
import { FormGroup, FormControl, Validators, FormBuilder, } from '@angular/forms'
import { GoodsReceiptItems } from "./goods-receipt-item.model"
import { ENUM_GRItemCategory } from "../../shared/shared-enums";
import * as moment from 'moment/moment';
export class GoodsReceipt {
  public GoodsReceiptValidator: FormGroup = null;
  public VendorName: string = "";
  public VendorNo: string = "";
  public IsCancel: boolean = false;
  AgingDays: number;
  SupplierId: number;
  VendorBillDate: string = moment().format('YYYY-MM-DD');
  GoodsReceiptDate: string = moment().format('YYYY-MM-DD');
  length: number;
  public FromDate: string = "";
  public ToDate: string = "";
  public CancelRemarks: string = "";
  //added for Verification
  public IsVerificationEnabled: boolean = false;
  public VerifierList: any[] = [];
  public VerifierIds: string;
  public CurrentVerificationLevel: number;
  public CurrentVerificationLevelCount: any;
  public MaxVerificationLevel: any;
  public IsVerificationAllowed: boolean;
  public VerificationStatus: string;
  public GRStatus: string;
  public DonationId: number;
  public ContactAddress: string = "";
  public MaterialCoaDate: string = "";/// Rajib  12/7/2020 tilaganga hospital
  public MaterialCoaNo: string;/// Rajib  12/7/2020 tilaganga hospital
  public PurchaseOrderDate: string = "";
  public IsSupplierApproved: boolean;
  public IsDeliveryTopClosed: boolean;
  public IsBoxNumbered: boolean;
  public StoreId: number;
  public GoodsReceiptID: number = 0;
  public GoodsArrivalNo: number = 0;
  public GoodsArrivalDate: string = moment().format('YYYY-MM-DD');
  public GoodsArrivalFiscalYearFormatter: string;
  public GoodsReceiptNo: number;
  public IMIRNo: number;
  public IMIRDate: string;
  public PurchaseOrderId: number = null;
  public TotalAmount: number = 0;
  public Remarks: string = "";
  public CreatedBy: number = 0;
  public CreatedOn: string = null;
  public VendorId: number = 0;
  public FiscalYearId: number;
  public CurrentFiscalYear: string = "";
  public BillNo: string = null;
  public ReceivedDate: string = moment().format('YYYY-MM-DD HH:MM:ss'); // Remove this date as it is not of any use.
  public ReceiptNo: string = null;
  public OrderDate: string = null;

  //SubTotal and VATTotal only for caculation and show purpose
  public SubTotal: number = 0;
  public VATTotal: number = 0;
  public CcCharge: number = 0;
  public Discount: number = 0;
  public DiscountAmount: number = 0;
  public TDSRate: number = 0;
  public TDSAmount: number = 0;
  public TotalWithTDS: number = 0;
  public PrintCount: number = 0;
  //public SelectedItem: any = null;//sud:11Apr'20-- this was misused.. hence removing.
  public CreditPeriod: number = 0;
  public PaymentMode: string;
  public OtherCharges: number = 0;
  public GoodsReceiptItem: Array<GoodsReceiptItems> = new Array<GoodsReceiptItems>();

  //for other charges
  public InsuranceCharge: number = 0;
  public CarriageFreightCharge: number = 0;
  public PackingCharge: number = 0;
  public TransportCourierCharge: number = 0;
  public OtherCharge: number = 0;

  //for edit option
  public IsTransferredToACC: boolean = false;
  public ModifiedBy: number = null;
  public ModifiedOn: Date = null;
  public IsDonation: boolean;
  GRGroupId: number;
  constructor() {

    var _formBuilder = new FormBuilder();
    this.GoodsReceiptValidator = _formBuilder.group({
      //sanjit: 2Apr'20: GoodsReceiptDate somehow throws validation error when use with danphe-date-picker, so it is commented. 
      'GoodsReceiptDate': [this.GoodsReceiptDate, Validators.compose([Validators.required])],
      'BillNo': ['', Validators.compose([Validators.required])],
      'VendorId': ['', Validators.compose([Validators.required])],
      'PaymentMode': ['', Validators.compose([Validators.required])],
    });
  }

  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.GoodsReceiptValidator.dirty;
    else
      return this.GoodsReceiptValidator.controls[fieldName].dirty;
  }


  public IsValid(): boolean { if (this.GoodsReceiptValidator.valid) { return true; } else { return false; } } public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined) {
      return this.GoodsReceiptValidator.valid;
    }
    else
      return !(this.GoodsReceiptValidator.hasError(validator, fieldName));
  }

  dateValidator(control: FormControl): { [key: string]: boolean } {
    //get current date, month and time
    var currDate = moment().format('YYYY-MM-DD');
    //if positive then selected date is of future else it of the past || selected year can't be of future
    if (control.value) {
      if ((moment(control.value).diff(currDate) > 0)
        || (moment(control.value).diff(currDate, 'years') > 1))
        return { 'wrongDate': true };
    }
    else
      return { 'wrongDate': true };

  }

  public updateItemDuplicationStatus() {
    if (this.GoodsReceiptItem) {
      for (var i = 0; i < this.GoodsReceiptItem.length; i++) {
        // for any this-item, if there exists another this-item with same ItemId but in different index, then we must set them as duplicates
        this.GoodsReceiptItem[i].isItemDuplicate = this.GoodsReceiptItem.some((thisitem, index) => thisitem.ItemId == this.GoodsReceiptItem[i].ItemId && index != i);
      }
    }
  }
  hasDuplicateItems(itemId?: number): boolean {
    return this.GoodsReceiptItem.some(a => a.isItemDuplicate == true);
  }
}

File: Code/Components/DanpheEMR.ServerModel/InventoryModels/GoodsReceiptModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel
{
    public class GoodsReceiptModel
    {
        [Key]
        public int GoodsReceiptID { get; set; }
        public bool? IsCancel { get; set; }
        public int GoodsArrivalNo { get; set; }
        public DateTime? VendorBillDate { get; set; }
        public DateTime? GoodsArrivalDate { get; set; }
        public int? IMIRNo { get; set; }
        public DateTime? IMIRDate { get; set; }
        public int? GoodsReceiptNo { get; set; }
        public DateTime? GoodsReceiptDate { get; set; }
        public int? PurchaseOrderId { get; set; }
        public string GRCategory { get; set; }
        public string Remarks { get; set; }
        public decimal TotalAmount { get; set; }
        public string CancelRemarks { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public int VendorId { get; set; }
        public string BillNo { get; set; }
        public DateTime? ReceivedDate { get; set; } // Remove this date as it is not of any use.
        public string ReceiptNo { get; set; }
        public DateTime? OrderDate { get; set; }

        public bool? IsTransferredToACC { get; set; }
        //SubTotal and VATTotal only for caculation and show purpose
        public decimal SubTotal { get; set; }
        public decimal VATTotal { get; set; }
        public decimal? TDSRate { get; set; }
        public decimal? TDSAmount { get; set; }
        public decimal? TotalWithTDS { get; set; }
        public decimal CcCharge { get; set; }
        public double Discount { get; set; }
        public decimal DiscountAmount { get; set; }
        public int PrintCount { get; set; }
        public int CreditPeriod { get; set; }
        public string PaymentMode { get; set; }
        public decimal? OtherCharges { get; set; }
        //for other charges
        public decimal? InsuranceCharge { get; set; }
        public decimal? CarriageFreightCharge { get; set; }
        public decimal? PackingCharge { get; set; }
        public decimal? TransportCourierCharge { get; set; }
        public decimal? OtherCharge { get; set; }
        public virtual List<GoodsReceiptItemsModel> GoodsReceiptItem { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public int? FiscalYearId { get; set; }
        [NotMapped]
        public string GoodsArrivalFiscalYearFormatted { get; set; }
        [NotMapped]
        public string FiscalYear { get; set; }

        //public VendorMasterModel Vendor { get; set; }
        //  public  decimal? TotalWithoutTDS { get; set; }

        [NotMapped]
        public int VoucherId { get; set; }
        [NotMapped]
        public string VoucherName { get; set; }
        [NotMapped]
        public string TransactionType { get; set; }
        [NotMapped]
        public decimal SalesAmount { get; set; }
        [NotMapped]
        public decimal VATAmount { get; set; }
        [NotMapped]
        public List<int> ReferenceIds { get; set; }
        [NotMapped]
        public string Type { get; set; }
        [NotMapped]
        public string VendorName { get; set; }
        [NotMapped]
        public int ItemId { get; set; }
        [NotMapped]
        public string ItemName { get; set; }
        [NotMapped]
        public List<SyncBillingAccountingModel> BillSyncs { get; set; }
        public bool IsVerificationEnabled { get; set; }
        public string VerifierIds { get; set; }
        [NotMapped]
        public List<POVerifier> VerifierList { get; set; }
        [NotMapped]
        public bool IsVerificationAllowed { get; set; }
        [NotMapped]
        public string VerificationStatus { get; set; }
        [NotMapped]
        public int CurrentVerificationLevel { get; set; }
        [NotMapped]
        public int CurrentVerificationLevelCount { get; set; }
        [NotMapped]
        public int MaxVerificationLevel { get; set; }
        public int? VerificationId { get; set; }
        public string GRStatus { get; set; }
        [NotMapped]
        public List<int> ReferenceIdsOne { get; set; }

        public int? CancelledBy { get; set; }//sud:15-Oct-2020: This property was there in db but not in here.
        public DateTime? CancelledOn { get; set; }//sud:15-Oct-2020: This property was there in db but not in here.
        public bool? IsDonation { get; set; }
        public int? DonationId { get; set; }


        public DateTime? MaterialCoaDate { get; set; }

        public string MaterialCoaNo { get; set; }

        public bool? IsSupplierApproved { get; set; }
        public bool? IsDeliveryTopClosed { get; set; }
        public bool? IsBoxNumbered { get; set; }
        public int? ReceivedBy { get; set; }
        public DateTime? ReceivedOn { get; set; }
        public string ReceivedRemarks { get; set; }
        public int StoreId { get; set; }

        /// <summary>
        /// Maintains sequence for GoodsReceiptNo
        /// </summary>
        public int? GRGroupId { get; set; }
        public bool IsPaymentDoneFromAcc { get; set; }
    }
}
